gdal/frmts/pcidsk/sdk/segment/cpcidskpolymodel.cpp

```diff
@@ -0,0 +1,317 @@
+ * PCI Geomatics, 90 Allstate Parkway, Markham, Ontario, Canada.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ ****************************************************************************/
+
+#include "pcidsk_poly.h"
+#include "segment/cpcidsksegment.h"
+#include "core/pcidsk_utils.h"
+#include "pcidsk_exception.h"
+#include "segment/cpcidskpolymodel.h"
+
+#include <vector>
+#include <string>
+#include <cassert>
+#include <cstring>
+
+using namespace PCIDSK;
+
+// Struct to store details of the RPC model
+struct CPCIDSKPolyModelSegment::PCIDSKPolyInfo
+{
+    // number of coefficients
+    unsigned int nNumCoeffs;
+
+    // pixels in the image
+    unsigned int nPixels;
+    // lines in the image
+    unsigned int nLines;
+
+    // Forward Coefficients (Geo2Img)
+    std::vector<double> vdfX1;
+    // Forward Coefficients (Geo2Img)
+    std::vector<double> vdfY1;
+    // Backward Coefficients Img2Geo
+    std::vector<double> vdfX2;
+    // Backward Coefficients Img2Geo
+    std::vector<double> vdfY2;
+
+    //map units of required projection
+    std::string oMapUnit;
+    //proj parm info of required projection
+    std::vector<double> oProjectionInfo;
+
+    // The raw segment data
+    PCIDSKBuffer seg_data;
+};
+
+CPCIDSKPolyModelSegment::CPCIDSKPolyModelSegment(PCIDSKFile *fileIn,
+                                                 int segmentIn,
+                                                 const char *segment_pointer) :
+    CPCIDSKSegment(fileIn, segmentIn, segment_pointer),
+    pimpl_(new CPCIDSKPolyModelSegment::PCIDSKPolyInfo),
+    loaded_(false),mbModified(false)
+{
+    Load();
+}
+
+
+CPCIDSKPolyModelSegment::~CPCIDSKPolyModelSegment()
+{
+    delete pimpl_;
+}
+
+// Load the contents of the segment
+void CPCIDSKPolyModelSegment::Load()
+{
+    // Check if we've already loaded the segment into memory
+    if (loaded_) {
+        return;
+    }
+
+    assert(data_size - 1024 == 7 * 512);
```

Comment:
could this assert fail on corrupted files ? If so it should be changed to a runtime check